### ISSUE TITLE
discuss launchers

### ISSUE THREAD

**Issue by andre-merzky** [ExaWorks/job-api-spec#16]
we need to discuss launchers (srun, mpirun, etc.); this is absent from the public API and from this document entirely, but chances are that we cannot entirely avoid this issue, so we might need to define an abstraction over various launchers. This could, in principle, be hidden in the adapters, but experience suggests that it's often (1) insufficient and (2) the reason why libraries like these are hard to work with (they never quite get the launching part right).
                                            
    - the public API need only reflect that the user could pick a non-default launcher by name                                                                                 
    - canonical implementation should have a nice way of abstracting over launchers

**Comment by hategan** [comment]
I'm going to volunteer some thoughts:

As @andre-merzky mentions, the public API probably only needs a `setLauncher(String)` somewhere, likely in the jobspec.

The basic idea of LRM connector is, in my mind this:
     - use jobspec to construct a submit script
     - use launcher to construct the actual launch command
         - launcher must have access to jobspec and allocation info (e.g., hostlist), which may be LRM/connector dependent
Consequently, the launcher essentially renders job and allocation info into a string that represents a command which is put into a submit script by a connector.

So it seems to me like the non-public part of a launcher can be summarized as `String getLaunchCommand(JobSpec spec)`

**Comment by andre-merzky** [comment]
> - use launcher to construct the actual launch command
> - launcher must have access to jobspec and allocation info (e.g., hostlist), which may be LRM/connector dependent
> Consequently, the launcher essentially renders job and allocation info into a string that represents a command which is put into a submit script by a connector.
> 
> So it seems to me like the non-public part of a launcher can be summarized as `String getLaunchCommand(JobSpec spec)`

I would suggest to keep this out of the document, or at least be very careful in keeping the implementation space open.  Launchers can be extremely divers - here is a selection of launcher approaches we handle or have worked with:

  - srun / aprun / mpirun / prun / ibrun / ...: create a command line as you suggest
  - RP to ORTE / PRTE: invoke a C-library call which sends a message
  - RP to Flux: invoke a python module call which interfaces to a C-library which sends a message
  - Flux internal: C method call or ZMQ message
  - Fireworks: place an task entry in a data base
  - RP / Raptor: reference a callable or an executable configured in a specific worker instance
  - Parsl: serialize a callable which is executed by a persistent worker on a node, or reference a script / command
  - ScaleMS: serialize a call context and data dependencies, send to execution worker

So, yes, some launchers may boil down to `getLaunchCommand()`, but that's really just one option of several...

Also, launchers do not always need to know the node list: often the backend can schedule and place the tasks.  Or in case for Raptor and Fireworks (and IIRC Parsl), tasks are pulled opportunistically from workers.

**Comment by hategan** [comment]
> launchers can be extremely divers

Agreed. However, in, say, PBS, wouldn't it have to be some commands in a submit script?

**Comment by andre-merzky** [comment]
Yes, in some backends it would be a script / command line.

**Comment by hategan** [comment]
> Yes, in some backends it would be a script / command line.

I think that's what is meant by "launcher". In the Flux case, I don't think there is a launcher. I.e., there is only one choice and that is to tell Flux to run things.

But then this points to launchers being rather specific to some connectors, and, therefore, an implementation detail. So perhaps, as you say, not much except the choice of a launcher belongs in the public API.

**Comment by hategan** [comment]
OK, so, for now, we add `get/setLauncher(String)` in the same place where, say, the queue is specified?

We also leave this issue open as a reminder that this isn't necessarily quite the end of the discussion?

**Comment by SteVwonder** [comment]
As part of discussing the launchers, we should also discuss the environment variables that certain launchers (e.g., launchers that support MPI) will set (e.g., `JPSI_RANK`).

**Comment by hategan** [comment]
Keeping this open in order to remember that we may need to document/prescribe some env vars that abstract over MPI env vars (such as PSIJ_RANK)
